### PR TITLE
Feature/ingress resources

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~~
 - [ ] <del>Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?</del>
 - [ ] <del>Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed? </del>
 - [ ] <del>Do comments answer the question of why design decisions were made? </del>

----

## Description of change

Introduce `ingress` resources in PodSpec `V2`;

## QA steps

- deploy CaaS charm with Ingress format like below;

```yaml
kubernetesResources:
  ingresses:
    - name: test-ingress
      labels:
        foo: bar
      annotations:
        ingress.kubernetes.io/rewrite-target: ""
        kubernetes.io/ingress.class: nginx
        kubernetes.io/ingress.allow-http: "true"
      spec:
        rules:
          - http:
              paths:
                - path: /
                  backend:
                    serviceName: test
                    servicePort: 80
```

- check result;

```console
$ kubectl get ingress.extensions/test-ingress -n <namespace> -o yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    ingress.kubernetes.io/rewrite-target: ""
    juju.io/controller: e34f5530-5850-4f32-8123-c36543aed23e
    juju.io/model: d6d8e529-dd1b-48b1-86aa-032185031702
    kubernetes.io/ingress.allow-http: "true"
    kubernetes.io/ingress.class: nginx
  creationTimestamp: "2019-11-27T07:03:58Z"
  generation: 1
  labels:
    foo: bar
    juju-app: mariadb-k8s
  name: test-ingress
  namespace: t1
  resourceVersion: "311230"
  selfLink: /apis/extensions/v1beta1/namespaces/t1/ingresses/test-ingress
  uid: 67983c83-f20f-4414-8722-948e9ef69ca7
spec:
  rules:
  - http:
      paths:
      - backend:
          serviceName: test
          servicePort: 80
        path: /
status:
  loadBalancer:
    ingress:
    - ip: 127.0.0.1
```

## Documentation changes

Yes

## Bug reference

No

### CODE REVIEW

**Comment by ycliuhw** [comment]
[Ingress Resource support](https://trello.com/c/RYjL4M58/982-ingress-resource-support)

File: caas/kubernetes/provider/specs/v2_test.go [hunk 1]
```diff
@@ -656,6 +701,34 @@ kubernetesResources:
...
+func (s *v2SpecsSuite) TestValidateIngresses(c *gc.C) {
+	specStr := versionHeader + `
+containers:
+  - name: gitlab-helper
+    image: gitlab-helper/latest
+    ports:
+    - containerPort: 8080
+      protocol: TCP
+kubernetesResources:
+  ingresses:
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/v2_test.go]: Plural seems a bit awkward. Perhaps "ingress". Or "ingressResources"

File: caas/kubernetes/provider/specs/v2.go [hunk 2]
```diff
@@ -98,6 +115,7 @@ type KubernetesResources struct {
 	CustomResources           map[string][]unstructured.Unstructured                       `json:"customResources,omitempty" yaml:"customResources,omitempty"`
 
 	ServiceAccounts []K8sServiceAccountSpec `json:"serviceAccounts,omitempty" yaml:"serviceAccounts,omitempty"`
+	Ingresses       []K8sIngressSpec        `json:"ingresses,omitempty" yaml:"ingresses,omitempty"`
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/v2.go]: "IngressResources" I think sounds better

File: caas/kubernetes/provider/ingress.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+func (k *kubernetesClient) deleteIngress(name string, uid k8stypes.UID) error {
+	err := k.client().ExtensionsV1beta1().Ingresses(k.namespace).Delete(name, newPreconditionDeleteOptions(uid))
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+	return errors.Trace(err)
+}
+
+func (k *kubernetesClient) listIngresses(labels map[string]string) ([]v1beta1.Ingress, error) {
```
Comment (wallyworld) [inline caas/kubernetes/provider/ingress.go]: IngressResources I think sounds better

File: caas/kubernetes/provider/ingress.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	}
+	_, err = k.updateIngress(spec)
+	return cleanUp, errors.Trace(err)
+}
+
+func (k *kubernetesClient) createIngress(ingress *v1beta1.Ingress) (*v1beta1.Ingress, error) {
+	purifyResource(ingress)
+	out, err := k.client().ExtensionsV1beta1().Ingresses(k.namespace).Create(ingress)
+	if k8serrors.IsAlreadyExists(err) {
+		return nil, errors.AlreadyExistsf("ingress %q", ingress.GetName())
```
Comment (wallyworld) [inline caas/kubernetes/provider/ingress.go]: "ingress resource %q"

etc

File: caas/kubernetes/provider/ingress.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
+// Copyright 2017 Canonical Ltd.
```
Comment (wallyworld) [inline caas/kubernetes/provider/ingress.go]: 2019

**Comment by ycliuhw** [comment]
[Ingress Resource support](https://trello.com/c/sFfMdzZK/4-ingress-resource-support)

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**